src/Sentry.Unity.Android/JniExecutor.cs

```diff
@@ -8,150 +9,157 @@ namespace Sentry.Unity.Android;
             }
         }
 
         AndroidJNI.DetachCurrentThread();
+        _logger?.LogDebug("JNI executor thread terminated");
     }
 
-    public TResult? Run<TResult>(Func<TResult?> jniOperation, TimeSpan? timeout = null)
+    private void ExecuteJniOperation(JniOperation operation)
     {
-        lock (_lock)
+        try
         {
-            timeout ??= DefaultTimeout;
-            using var timeoutCts = new CancellationTokenSource(timeout.Value);
-            _taskCompletionSource = new TaskCompletionSource<object?>();
-            _currentTask = jniOperation;
-            _taskEvent.Set();
-
-            try
-            {
-                _taskCompletionSource.Task.Wait(timeoutCts.Token);
-                return (TResult?)_taskCompletionSource.Task.Result;
-            }
-            catch (OperationCanceledException)
+            switch (operation.Action)
             {
-                _logger?.LogError("JNI execution timed out.");
-                return default;
+                case Action action:
+                    action.Invoke();
+                    operation.CompletionSource?.TrySetResult(null);
+                    break;
+                case Func<bool> func1:
+                    var result1 = func1.Invoke();
+                    operation.CompletionSource?.TrySetResult(result1);
+                    break;
+                case Func<bool?> func2:
+                    var result2 = func2.Invoke();
+                    operation.CompletionSource?.TrySetResult(result2);
+                    break;
+                case Func<string?> func3:
+                    var result3 = func3.Invoke();
+                    operation.CompletionSource?.TrySetResult(result3);
+                    break;
+                default:
+                    var error = new NotImplementedException($"Task type '{operation.Action.GetType()}' is not implemented in JniExecutor");
+                    _logger?.LogError(error, "Unsupported JNI operation type");
+                    operation.CompletionSource?.TrySetException(error);
+                    break;
             }
-            catch (Exception e)
-            {
-                _logger?.LogError(e, "Error during JNI execution.");
-                return default;
-            }
-            finally
+        }
+        catch (Exception ex)
+        {
+            _logger?.LogError(ex, "Error during JNI operation execution.");
+            operation.CompletionSource?.TrySetException(ex);
+        }
+    }
+
+    /// <summary>
+    /// Runs a JNI operation that returns a result, waiting for completion with a timeout
+    /// </summary>
+    public TResult? Run<TResult>(Func<TResult?> jniOperation, TimeSpan? timeout = null)
+    {
+        timeout ??= DefaultTimeout;
+        var completionSource = new TaskCompletionSource<object?>();
+        var operation = new JniOperation(jniOperation, completionSource);
+
+        _operationQueue.Enqueue(operation);
```

Comment:
what's the queue capacity? if it's full, it blocks here too

Reply:
Fixed.